// @fuel-ts/providers@0.16.0/dist/index.js
"use strict";var Qt=Object.create;var H=Object.defineProperty;var Lt=Object.getOwnPropertyDescriptor;var Dt=Object.getOwnPropertyNames;var Ot=Object.getPrototypeOf,$t=Object.prototype.hasOwnProperty;var Ft=(r,t)=>{for(var e in t)H(r,e,{get:t[e],enumerable:!0})},lt=(r,t,e,a)=>{if(t&&typeof t=="object"||typeof t=="function")for(let n of Dt(t))!$t.call(r,n)&&n!==e&&H(r,n,{get:()=>t[n],enumerable:!(a=Lt(t,n))||a.enumerable});return r};var yt=(r,t,e)=>(e=r!=null?Qt(Ot(r)):{},lt(t||!r||!r.__esModule?H(e,"default",{value:r,enumerable:!0}):e,r)),Ut=r=>lt(H({},"__esModule",{value:!0}),r);var Et=(r,t,e)=>{if(!t.has(r))throw TypeError("Cannot "+e)};var dt=(r,t,e)=>{if(t.has(r))throw TypeError("Cannot add the same private member more than once");t instanceof WeakSet?t.add(r):t.set(r,e)};var mt=(r,t,e)=>(Et(r,t,"access private method"),e);var Ie={};Ft(Ie,{ChangeOutputCollisionError:()=>Z,CoinStatus:()=>nt,CreateTransactionRequest:()=>O,LogReader:()=>wt,NoWitnessAtIndexError:()=>z,NoWitnessByOwnerError:()=>X,Provider:()=>E,ScriptTransactionRequest:()=>R,TransactionResponse:()=>U,TransactionType:()=>i.TransactionType,calculatePriceWithFactor:()=>Q,coinQuantityfy:()=>_,getGasUsedFromReceipts:()=>B,getSignableTransaction:()=>ft,getTransactionId:()=>ye,inputify:()=>ot,outputify:()=>it,returnZeroScript:()=>ct,transactionRequestify:()=>$});module.exports=Ut(Ie);var gt=require("@ethersproject/bytes"),rt=require("@fuel-ts/constants"),St=require("@fuel-ts/math"),_=r=>{var a,n;let t,e;return Array.isArray(r)?(e=r[0],t=(a=r[1])!=null?a:rt.NativeAssetId):(e=r.amount,t=(n=r.assetId)!=null?n:rt.NativeAssetId),{assetId:(0,gt.hexlify)(t),amount:(0,St.bn)(e)}};var u=yt(require("graphql-tag")),nt=(e=>(e.Spent="SPENT",e.Unspent="UNSPENT",e))(nt||{});var k=u.default`
  fragment transactionFragment on Transaction {
    id
    rawPayload
    status {
      type: __typename
      ... on SubmittedStatus {
        time
      }
      ... on SuccessStatus {
        block {
          id
        }
        time
        programState {
          returnType
          data
        }
      }
      ... on FailureStatus {
        block {
          id
        }
        time
        reason
      }
    }
  }
`,It=u.default`
  fragment receiptFragment on Receipt {
    data
    rawPayload
  }
`,st=u.default`
  fragment coinFragment on Coin {
    utxoId
    owner
    amount
    assetId
    maturity
    status
    blockCreated
  }
`,qt=u.default`
  fragment balanceFragment on Balance {
    owner
    amount
    assetId
  }
`,W=u.default`
  fragment blockFragment on Block {
    id
    height
    producer
    transactions {
      id
    }
    time
  }
`,Vt=u.default`
  fragment chainInfoFragment on ChainInfo {
    name
    baseChainHeight
    peerCount
    consensusParameters {
      gasPriceFactor
      maxGasPerTx
      maxGasPerTx
      maxScriptLength
    }
    latestBlock {
      ...blockFragment
    }
  }
  ${W}
`,Nt=u.default`
  query getVersion {
    nodeInfo {
      nodeVersion
    }
  }
`,Ht=u.default`
  query getInfo {
    nodeInfo {
      nodeVersion
      minGasPrice
    }
  }
`,Wt=u.default`
  query getChain {
    chain {
      ...chainInfoFragment
    }
  }
  ${Vt}
`,vt=u.default`
  query getTransaction($transactionId: TransactionId!) {
    transaction(id: $transactionId) {
      ...transactionFragment
    }
  }
  ${k}
`,Kt=u.default`
  query getTransactionWithReceipts($transactionId: TransactionId!) {
    transaction(id: $transactionId) {
      ...transactionFragment
      receipts {
        ...receiptFragment
      }
    }
  }
  ${k}
  ${It}
`,Zt=u.default`
  query getTransactions($after: String, $before: String, $first: Int, $last: Int) {
    transactions(after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...transactionFragment
        }
      }
    }
  }
  ${k}
`,zt=u.default`
  query getTransactionsByOwner(
    $owner: Address!
    $after: String
    $before: String
    $first: Int
    $last: Int
  ) {
    transactionsByOwner(owner: $owner, after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...transactionFragment
        }
      }
    }
  }
  ${k}
`,Xt=u.default`
  query getBlock($blockId: BlockId, $blockHeight: U64) {
    block(id: $blockId, height: $blockHeight) {
      ...blockFragment
    }
  }
  ${W}
`,Jt=u.default`
  query getBlockWithTransactions($blockId: BlockId, $blockHeight: U64) {
    block(id: $blockId, height: $blockHeight) {
      ...blockFragment
      transactions {
        ...transactionFragment
      }
    }
  }
  ${W}
  ${k}
`,jt=u.default`
  query getBlocks($after: String, $before: String, $first: Int, $last: Int) {
    blocks(after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...blockFragment
        }
      }
    }
  }
  ${W}
`,Yt=u.default`
  query getCoin($coinId: UtxoId!) {
    coin(utxoId: $coinId) {
      ...coinFragment
    }
  }
  ${st}
`,te=u.default`
  query getCoins(
    $filter: CoinFilterInput!
    $after: String
    $before: String
    $first: Int
    $last: Int
  ) {
    coins(filter: $filter, after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...coinFragment
        }
      }
    }
  }
  ${st}
`,ee=u.default`
  query getCoinsToSpend(
    $owner: Address!
    $spendQuery: [SpendQueryElementInput!]!
    $maxInputs: Int
    $excludedIds: [UtxoId!]
  ) {
    coinsToSpend(
      owner: $owner
      spendQuery: $spendQuery
      maxInputs: $maxInputs
      excludedIds: $excludedIds
    ) {
      ...coinFragment
    }
  }
  ${st}
`,ae=u.default`
  query getContract($contractId: ContractId!) {
    contract(id: $contractId) {
      bytecode
      id
    }
  }
`,re=u.default`
  query getBalance($owner: Address!, $assetId: AssetId!) {
    balance(owner: $owner, assetId: $assetId) {
      ...balanceFragment
    }
  }
  ${qt}
`,ne=u.default`
  query getBalances(
    $filter: BalanceFilterInput!
    $after: String
    $before: String
    $first: Int
    $last: Int
  ) {
    balances(filter: $filter, after: $after, before: $before, first: $first, last: $last) {
      edges {
        node {
          ...balanceFragment
        }
      }
    }
  }
  ${qt}
`,se=u.default`
  mutation dryRun($encodedTransaction: HexString!, $utxoValidation: Boolean) {
    dryRun(tx: $encodedTransaction, utxoValidation: $utxoValidation) {
      ...receiptFragment
    }
  }
  ${It}
`,oe=u.default`
  mutation submit($encodedTransaction: HexString!) {
    submit(tx: $encodedTransaction) {
      id
    }
  }
`,ie=u.default`
  mutation startSession {
    startSession
  }
`,ce=u.default`
  mutation endSession($sessionId: ID!) {
    endSession(id: $sessionId)
  }
`,ue=u.default`
  mutation execute($sessionId: ID!, $op: String!) {
    execute(id: $sessionId, op: $op)
  }
`,pe=u.default`
  mutation reset($sessionId: ID!) {
    reset(id: $sessionId)
  }
`,le=(r,t,e)=>r();function bt(r,t=le){return{getVersion(e,a){return t(n=>r.request(Nt,e,{...a,...n}),"getVersion","query")},getInfo(e,a){return t(n=>r.request(Ht,e,{...a,...n}),"getInfo","query")},getChain(e,a){return t(n=>r.request(Wt,e,{...a,...n}),"getChain","query")},getTransaction(e,a){return t(n=>r.request(vt,e,{...a,...n}),"getTransaction","query")},getTransactionWithReceipts(e,a){return t(n=>r.request(Kt,e,{...a,...n}),"getTransactionWithReceipts","query")},getTransactions(e,a){return t(n=>r.request(Zt,e,{...a,...n}),"getTransactions","query")},getTransactionsByOwner(e,a){return t(n=>r.request(zt,e,{...a,...n}),"getTransactionsByOwner","query")},getBlock(e,a){return t(n=>r.request(Xt,e,{...a,...n}),"getBlock","query")},getBlockWithTransactions(e,a){return t(n=>r.request(Jt,e,{...a,...n}),"getBlockWithTransactions","query")},getBlocks(e,a){return t(n=>r.request(jt,e,{...a,...n}),"getBlocks","query")},getCoin(e,a){return t(n=>r.request(Yt,e,{...a,...n}),"getCoin","query")},getCoins(e,a){return t(n=>r.request(te,e,{...a,...n}),"getCoins","query")},getCoinsToSpend(e,a){return t(n=>r.request(ee,e,{...a,...n}),"getCoinsToSpend","query")},getContract(e,a){return t(n=>r.request(ae,e,{...a,...n}),"getContract","query")},getBalance(e,a){return t(n=>r.request(re,e,{...a,...n}),"getBalance","query")},getBalances(e,a){return t(n=>r.request(ne,e,{...a,...n}),"getBalances","query")},dryRun(e,a){return t(n=>r.request(se,e,{...a,...n}),"dryRun","mutation")},submit(e,a){return t(n=>r.request(oe,e,{...a,...n}),"submit","mutation")},startSession(e,a){return t(n=>r.request(ie,e,{...a,...n}),"startSession","mutation")},endSession(e,a){return t(n=>r.request(ce,e,{...a,...n}),"endSession","mutation")},execute(e,a){return t(n=>r.request(ue,e,{...a,...n}),"execute","mutation")},reset(e,a){return t(n=>r.request(pe,e,{...a,...n}),"reset","mutation")}}}var q=require("@ethersproject/bytes"),_t=require("@fuel-ts/abi-coder"),ut=require("@fuel-ts/constants"),o=require("@fuel-ts/math"),S=require("@fuel-ts/transactions"),kt=require("graphql-request"),At=yt(require("lodash.clonedeep"));var l=require("@ethersproject/bytes"),v=require("@fuel-ts/constants"),C=require("@fuel-ts/math"),A=require("@fuel-ts/transactions"),ot=r=>{var t,e,a;switch(r.type){case A.InputType.Coin:{let n=(0,l.arrayify)((t=r.predicate)!=null?t:"0x"),s=(0,l.arrayify)((e=r.predicateData)!=null?e:"0x");return{type:A.InputType.Coin,utxoID:{transactionId:(0,l.hexlify)((0,l.arrayify)(r.id).slice(0,32)),outputIndex:(0,l.arrayify)(r.id)[32]},owner:(0,l.hexlify)(r.owner),amount:(0,C.bn)(r.amount),assetId:(0,l.hexlify)(r.assetId),txPointer:{blockHeight:(0,C.toNumber)((0,l.arrayify)(r.txPointer).slice(0,8)),txIndex:(0,C.toNumber)((0,l.arrayify)(r.txPointer).slice(8,16))},witnessIndex:r.witnessIndex,maturity:(a=r.maturity)!=null?a:0,predicateLength:n.length,predicateDataLength:s.length,predicate:(0,l.hexlify)(n),predicateData:(0,l.hexlify)(s)}}case A.InputType.Contract:return{type:A.InputType.Contract,utxoID:{transactionId:v.ZeroBytes32,outputIndex:0},balanceRoot:v.ZeroBytes32,stateRoot:v.ZeroBytes32,txPointer:{blockHeight:(0,C.toNumber)((0,l.arrayify)(r.txPointer).slice(0,8)),txIndex:(0,C.toNumber)((0,l.arrayify)(r.txPointer).slice(8,16))},contractID:(0,l.hexlify)(r.contractId)};default:throw new Error("Invalid Input type")}};var f=require("@ethersproject/bytes"),P=require("@fuel-ts/constants"),M=require("@fuel-ts/math"),g=require("@fuel-ts/transactions"),it=r=>{switch(r.type){case g.OutputType.Coin:return{type:g.OutputType.Coin,to:(0,f.hexlify)(r.to),amount:(0,M.bn)(r.amount),assetId:(0,f.hexlify)(r.assetId)};case g.OutputType.Contract:return{type:g.OutputType.Contract,inputIndex:r.inputIndex,balanceRoot:P.ZeroBytes32,stateRoot:P.ZeroBytes32};case g.OutputType.Withdrawal:return{type:g.OutputType.Withdrawal,to:(0,f.hexlify)(r.to),amount:(0,M.bn)(r.amount),assetId:(0,f.hexlify)(r.assetId)};case g.OutputType.Change:return{type:g.OutputType.Change,to:(0,f.hexlify)(r.to),amount:(0,M.bn)(0),assetId:(0,f.hexlify)(r.assetId)};case g.OutputType.Variable:return{type:g.OutputType.Variable,to:P.ZeroBytes32,amount:(0,M.bn)(0),assetId:P.ZeroBytes32};case g.OutputType.ContractCreated:return{type:g.OutputType.ContractCreated,contractId:(0,f.hexlify)(r.contractId),stateRoot:(0,f.hexlify)(r.stateRoot)};default:throw new Error("Invalid Output type")}};var m=require("@ethersproject/bytes"),T=require("@fuel-ts/address"),h=require("@fuel-ts/constants"),D=require("@fuel-ts/math"),i=require("@fuel-ts/transactions");var xt=require("@ethersproject/sha2"),G=require("@fuel-ts/constants"),w=require("@fuel-ts/math"),d=require("@fuel-ts/transactions"),ft=r=>{let t={...r};switch(t.type){case d.TransactionType.Script:{t.receiptsRoot=G.ZeroBytes32;break}case d.TransactionType.Create:break;default:throw new Error("Not implemented")}return t.inputs=t.inputs.map(e=>e.type===d.InputType.Contract?{...e,utxoID:{transactionId:G.ZeroBytes32,outputIndex:0},balanceRoot:G.ZeroBytes32,stateRoot:G.ZeroBytes32}:e),t.outputs=t.outputs.map(e=>{switch(e.type){case d.OutputType.Contract:return{...e,balanceRoot:G.ZeroBytes32,stateRoot:G.ZeroBytes32};case d.OutputType.Change:return{...e,amount:(0,w.bn)(0)};case d.OutputType.Variable:return{...e,to:G.ZeroBytes32,amount:(0,w.bn)(0),assetId:G.ZeroBytes32};default:return e}}),t},ye=r=>{let t=ft(r),e=new d.TransactionCoder().encode(t);return(0,xt.sha256)(e)},Q=(r,t,e)=>(0,w.bn)(Math.ceil(r.toNumber()/e.toNumber())*t.toNumber()),B=r=>{let t=r.find(e=>e.type===d.ReceiptType.ScriptResult);return t&&t.type===d.ReceiptType.ScriptResult?t.gasUsed:(0,w.bn)(0)};var L=require("@ethersproject/bytes"),de=r=>{let t=new Uint8Array(32);return t.set((0,L.arrayify)(r)),t},Gt=r=>{let t,e;return Array.isArray(r)?(t=r[0],e=r[1]):(t=r.key,e=r.value),{key:(0,L.hexlify)(t),value:(0,L.hexlify)(de(e))}};var K=require("@ethersproject/bytes"),Tt=r=>{let t=(0,K.arrayify)(r);return{data:(0,K.hexlify)(t),dataLength:t.length}};var ct={bytes:(0,m.arrayify)("0x24000000"),encodeScriptData:()=>new Uint8Array(0)},Z=class extends Error{constructor(){super(...arguments);this.name="ChangeOutputCollisionError";this.message='A ChangeOutput with the same "assetId" already exists for a different "to" address'}},z=class extends Error{constructor(e){super();this.index=e;this.name="NoWitnessAtIndexError";this.message=`Witness at index "${e}" was not found`}},X=class extends Error{constructor(e){super();this.owner=e;this.name="NoWitnessByOwnerError";this.message=`A witness for the given owner "${e}" was not found`}},J=class{constructor({gasPrice:t,gasLimit:e,maturity:a,inputs:n,outputs:s,witnesses:c}={}){this.inputs=[];this.outputs=[];this.witnesses=[];this.gasPrice=(0,D.bn)(t!=null?t:0),this.gasLimit=(0,D.bn)(e!=null?e:0),this.maturity=a!=null?a:0,this.inputs=[...n!=null?n:[]],this.outputs=[...s!=null?s:[]],this.witnesses=[...c!=null?c:[]]}getBaseTransaction(){var n,s,c,p,x,b;let t=(s=(n=this.inputs)==null?void 0:n.map(ot))!=null?s:[],e=(p=(c=this.outputs)==null?void 0:c.map(it))!=null?p:[],a=(b=(x=this.witnesses)==null?void 0:x.map(Tt))!=null?b:[];return{gasPrice:this.gasPrice,gasLimit:this.gasLimit,maturity:this.maturity,inputs:t,outputs:e,witnesses:a,inputsCount:t.length,outputsCount:e.length,witnessesCount:a.length}}toTransactionBytes(){return new i.TransactionCoder().encode(this.toTransaction())}pushInput(t){return this.inputs.push(t),this.inputs.length-1}pushOutput(t){return this.outputs.push(t),this.outputs.length-1}createWitness(){return this.witnesses.push("0x"),this.witnesses.length-1}updateWitness(t,e){if(!this.witnesses[t])throw new z(t);this.witnesses[t]=e}getCoinInputs(){return this.inputs.filter(t=>t.type===i.InputType.Coin)}getCoinOutputs(){return this.outputs.filter(t=>t.type===i.OutputType.Coin)}getChangeOutputs(){return this.outputs.filter(t=>t.type===i.OutputType.Change)}getCoinInputWitnessIndexByOwner(t){var a,n;let e=(0,T.addressify)(t);return(n=(a=this.inputs.find(s=>s.type===i.InputType.Coin&&s.owner===e.toB256()))==null?void 0:a.witnessIndex)!=null?n:null}updateWitnessByCoinInputOwner(t,e){let a=this.getCoinInputWitnessIndexByOwner(t);if(!a)throw new X((0,T.addressify)(t));this.updateWitness(a,e)}addCoin(t){let e=this.getCoinInputWitnessIndexByOwner(T.Address.fromB256(t.owner));typeof e!="number"&&(e=this.createWitness()),this.pushInput({type:i.InputType.Coin,...t,witnessIndex:e,txPointer:"0x00000000000000000000000000000000"});let a=this.getChangeOutputs().find(n=>(0,m.hexlify)(n.assetId)===t.assetId);if(a&&(0,m.hexlify)(a.to)!==t.owner)throw new Z;a||this.pushOutput({type:i.OutputType.Change,to:t.owner,assetId:t.assetId})}addCoins(t){t.forEach(e=>this.addCoin(e))}addCoinOutput(t,e,a=h.NativeAssetId){this.pushOutput({type:i.OutputType.Coin,to:(0,T.addressify)(t).toB256(),amount:e,assetId:a})}addCoinOutputs(t,e){e.map(_).forEach(a=>{this.pushOutput({type:i.OutputType.Coin,to:(0,T.addressify)(t).toB256(),amount:a.amount,assetId:a.assetId})})}byteSize(){return this.toTransactionBytes().length}chargeableByteSize(){let t=this.witnesses.reduce((e,a)=>e+(0,m.arrayify)(a).length,0);return(0,D.bn)(this.toTransactionBytes().length-t)}calculateFee(){let t=Q(this.gasLimit,this.gasPrice,i.GAS_PRICE_FACTOR);return{assetId:h.NativeAssetId,amount:t.isZero()?(0,D.bn)(1):t}}},R=class extends J{constructor({script:e,scriptData:a,...n}={}){super(n);this.type=i.TransactionType.Script;this.script=(0,m.arrayify)(e!=null?e:ct.bytes),this.scriptData=(0,m.arrayify)(a!=null?a:ct.encodeScriptData())}static from(e){return e instanceof this?e:new this(e)}toTransaction(){var n,s;let e=(0,m.arrayify)((n=this.script)!=null?n:"0x"),a=(0,m.arrayify)((s=this.scriptData)!=null?s:"0x");return{type:i.TransactionType.Script,...super.getBaseTransaction(),scriptLength:e.length,scriptDataLength:a.length,receiptsRoot:h.ZeroBytes32,script:(0,m.hexlify)(e),scriptData:(0,m.hexlify)(a)}}getContractInputs(){return this.inputs.filter(e=>e.type===i.InputType.Contract)}getContractOutputs(){return this.outputs.filter(e=>e.type===i.OutputType.Contract)}getVariableOutputs(){return this.outputs.filter(e=>e.type===i.OutputType.Variable)}setScript(e,a){this.script=e.bytes,this.scriptData=e.encodeScriptData(a),this.bytesOffset===void 0&&(this.bytesOffset=this.scriptData.byteLength)}addVariableOutputs(e=1){let a=e;for(;a;)this.pushOutput({type:i.OutputType.Variable}),a-=1;return this.outputs.length-1}addContract(e){let a=(0,T.addressify)(e);if(this.getContractInputs().find(s=>s.contractId===a.toB256()))return;let n=super.pushInput({type:i.InputType.Contract,contractId:a.toB256(),txPointer:"0x00000000000000000000000000000000"});this.pushOutput({type:i.OutputType.Contract,inputIndex:n})}},O=class extends J{constructor({bytecodeWitnessIndex:e,salt:a,storageSlots:n,...s}={}){super(s);this.type=i.TransactionType.Create;this.bytecodeWitnessIndex=e!=null?e:0,this.salt=(0,m.hexlify)(a!=null?a:h.ZeroBytes32),this.storageSlots=[...n!=null?n:[]]}static from(e){return e instanceof this?e:new this(e)}toTransaction(){var s,c;let e=this.getBaseTransaction(),a=this.bytecodeWitnessIndex,n=(c=(s=this.storageSlots)==null?void 0:s.map(Gt))!=null?c:[];return{type:i.TransactionType.Create,...e,bytecodeLength:e.witnesses[a].dataLength/4,bytecodeWitnessIndex:a,storageSlotsCount:n.length,salt:this.salt?(0,m.hexlify)(this.salt):h.ZeroBytes32,storageSlots:n}}getContractCreatedOutputs(){return this.outputs.filter(e=>e.type===i.OutputType.ContractCreated)}addContractCreatedOutput(e,a){this.pushOutput({type:i.OutputType.ContractCreated,contractId:e,stateRoot:a})}},$=r=>{if(r instanceof R||r instanceof O)return r;switch(r.type){case i.TransactionType.Script:return R.from(r);case i.TransactionType.Create:return O.from(r);default:throw new Error(`Unknown transaction type: ${r.type}`)}};var ht=require("@ethersproject/bytes"),Rt=require("@fuel-ts/math"),F=require("@fuel-ts/transactions");var Ct=r=>{let t=new F.ReceiptCoder().decode((0,ht.arrayify)(r.rawPayload),0)[0];switch(t.type){case F.ReceiptType.ReturnData:return{...t,data:r.data};case F.ReceiptType.LogData:return{...t,data:r.data};default:return t}},j,Bt,U=class{constructor(t,e,a){dt(this,j);this.gasUsed=(0,Rt.bn)(0);this.id=t,this.request=e,this.provider=a}async waitForResult(){var e;let t=await mt(this,j,Bt).call(this);switch((e=t.status)==null?void 0:e.type){case"SubmittedStatus":throw new Error("Not yet implemented");case"FailureStatus":{let a=t.receipts.map(Ct);return this.gasUsed=B(a),{status:{type:"failure",reason:t.status.reason},receipts:a,transactionId:this.id,blockId:t.status.block.id,time:t.status.time}}case"SuccessStatus":{let a=t.receipts.map(Ct);return this.gasUsed=B(a),{status:{type:"success",programState:t.status.programState},receipts:a,transactionId:this.id,blockId:t.status.block.id,time:t.status.time}}default:throw new Error("Invalid Transaction status")}}async wait(){let t=await this.waitForResult();if(t.status.type==="failure")throw new Error(`Transaction failed: ${t.status.reason}`);return t}};j=new WeakSet,Bt=async function(){let{transaction:t}=await this.provider.operations.getTransactionWithReceipts({transactionId:this.id});if(!t)throw new Error("No Transaction was received from the client.");return t};var me=r=>{let t=new S.ReceiptCoder().decode((0,q.arrayify)(r.rawPayload),0)[0];switch(t.type){case S.ReceiptType.ReturnData:return{...t,data:r.data};case S.ReceiptType.LogData:return{...t,data:r.data};default:return t}},ge=r=>({name:r.name,baseChainHeight:(0,o.bn)(r.baseChainHeight),peerCount:r.peerCount,consensusParameters:{gasPriceFactor:(0,o.bn)(r.consensusParameters.gasPriceFactor),maxGasPerTx:(0,o.bn)(r.consensusParameters.maxGasPerTx),maxScriptLength:(0,o.bn)(r.consensusParameters.maxScriptLength)},latestBlock:{id:r.latestBlock.id,height:(0,o.bn)(r.latestBlock.height),producer:r.latestBlock.producer,time:r.latestBlock.time,transactions:r.latestBlock.transactions.map(t=>({id:t.id}))}}),Se=r=>({minGasPrice:(0,o.bn)(r.minGasPrice),nodeVersion:r.nodeVersion}),E=class{constructor(t){this.url=t;let e=new kt.GraphQLClient(t);this.operations=bt(e)}async getVersion(){let{nodeInfo:{nodeVersion:t}}=await this.operations.getVersion();return t}async getNetwork(){return{name:"fuelv2",chainId:3735928559}}async getBlockNumber(){let{chain:t}=await this.operations.getChain();return(0,o.bn)(t.latestBlock.height,10)}async getNodeInfo(){let{nodeInfo:t}=await this.operations.getInfo();return Se(t)}async getChain(){let{chain:t}=await this.operations.getChain();return ge(t)}async sendTransaction(t){let e=$(t),a=(0,q.hexlify)(e.toTransactionBytes()),{gasUsed:n,minGasPrice:s}=await this.getTransactionCost(e,0);if((0,o.bn)(n).gt((0,o.bn)(e.gasLimit)))throw new Error(`gasLimit(${e.gasLimit}) is lower than the required (${n})`);if((0,o.bn)(s).gt((0,o.bn)(e.gasPrice)))throw new Error(`gasPrice(${e.gasPrice}) is lower than the required ${s}`);let{submit:{id:c}}=await this.operations.submit({encodedTransaction:a});return new U(c,e,this)}async call(t,{utxoValidation:e}={}){let a=$(t),n=(0,q.hexlify)(a.toTransactionBytes()),{dryRun:s}=await this.operations.dryRun({encodedTransaction:n,utxoValidation:e||!1});return{receipts:s.map(me)}}async getTransactionCost(t,e=.2){let a=$((0,At.default)(t)),{minGasPrice:n}=await this.getNodeInfo(),s=(0,o.max)(a.gasPrice,n),c=1+e;a.gasLimit=S.MAX_GAS_PER_TX,a.gasPrice=(0,o.bn)(0);let{receipts:p}=await this.call(a),x=(0,o.multiply)(B(p),c),b=Q(x,s,S.GAS_PRICE_FACTOR);return{minGasPrice:n,gasPrice:s,gasUsed:x,fee:b}}async getCoins(t,e,a){return(await this.operations.getCoins({first:10,...a,filter:{owner:t.toB256(),assetId:e&&(0,q.hexlify)(e)}})).coins.edges.map(c=>c.node).map(c=>({id:c.utxoId,assetId:c.assetId,amount:(0,o.bn)(c.amount),owner:c.owner,status:c.status,maturity:(0,o.bn)(c.maturity).toNumber(),blockCreated:(0,o.bn)(c.blockCreated)}))}async getCoinsToSpend(t,e,a,n){return(await this.operations.getCoinsToSpend({owner:t.toB256(),spendQuery:e.map(_).map(p=>({assetId:(0,q.hexlify)(p.assetId),amount:p.amount.toString(10)})),maxInputs:a,excludedIds:n==null?void 0:n.map(p=>(0,q.hexlify)(p))})).coinsToSpend.map(p=>({id:p.utxoId,status:p.status,assetId:p.assetId,amount:(0,o.bn)(p.amount),owner:p.owner,maturity:(0,o.bn)(p.maturity).toNumber(),blockCreated:(0,o.bn)(p.blockCreated)}))}async getBlock(t){let e;typeof t=="number"?e={blockHeight:(0,o.bn)(t).toString(10)}:t==="latest"?e={blockHeight:(await this.getBlockNumber()).toString(10)}:e={blockId:(0,o.bn)(t).toString(10)};let{block:a}=await this.operations.getBlock(e);return a?{id:a.id,height:(0,o.bn)(a.height),time:a.time,producer:a.producer,transactionIds:a.transactions.map(n=>n.id)}:null}async getBlockWithTransactions(t){let e;typeof t=="number"?e={blockHeight:(0,o.bn)(t).toString(10)}:t==="latest"?e={blockHeight:(await this.getBlockNumber()).toString()}:e={blockId:t};let{block:a}=await this.operations.getBlockWithTransactions(e);return a?{id:a.id,height:(0,o.bn)(a.height,10),time:a.time,producer:a.producer,transactionIds:a.transactions.map(n=>n.id),transactions:a.transactions.map(n=>{var s;return(s=new S.TransactionCoder().decode((0,q.arrayify)(n.rawPayload),0))==null?void 0:s[0]})}:null}async getTransaction(t){var a;let{transaction:e}=await this.operations.getTransaction({transactionId:t});return e?(a=new S.TransactionCoder().decode((0,q.arrayify)(e.rawPayload),0))==null?void 0:a[0]:null}async getContract(t){let{contract:e}=await this.operations.getContract({contractId:t});return e||null}async getBalance(t,e){let{balance:a}=await this.operations.getBalance({owner:t.toB256(),assetId:(0,q.hexlify)(e)});return(0,o.bn)(a.amount,10)}async getBalances(t,e){return(await this.operations.getBalances({first:10,...e,filter:{owner:t.toB256()}})).balances.edges.map(s=>s.node).map(s=>({assetId:s.assetId,amount:(0,o.bn)(s.amount)}))}async buildSpendPredicate(t,e,a,n,s=ut.NativeAssetId,c,p){let x=await this.getCoinsToSpend(t.address,[[e,s]]),b={fundTransaction:!0,...c},I=new R({gasLimit:S.MAX_GAS_PER_TX,...b}),tt;n&&t.types&&(tt=new _t.AbiCoder().encode(t.types,n));let et=x.reduce((N,pt)=>(I.addCoin({...pt,predicate:t.bytes,predicateData:tt}),I.outputs=[],N.add(pt.amount)),(0,o.bn)(0));I.addCoinOutput(a,et,s);let at=[];if(b.fundTransaction&&at.push(I.calculateFee()),at.length&&p){let N=await this.getCoinsToSpend(p,at);I.addCoins(N)}return I}async submitSpendPredicate(t,e,a,n,s=ut.NativeAssetId,c,p){var b;let x=await this.buildSpendPredicate(t,e,a,n,s,c,p);try{return await(await this.sendTransaction(x)).waitForResult()}catch(I){throw(((b=I==null?void 0:I.response)==null?void 0:b.errors)||[]).some(({message:et})=>et.includes("unexpected block execution error TransactionValidity(InvalidPredicate"))?new Error("Invalid Predicate"):I}}};var Pt=require("@ethersproject/bytes"),Mt=require("@fuel-ts/abi-coder"),Y=require("@fuel-ts/transactions"),V=class{constructor(t){this.logs=t.filter(({type:e})=>e===Y.ReceiptType.Log||e===Y.ReceiptType.LogData)}toArray(){return this.logs.map(t=>{if(t.type===Y.ReceiptType.LogData){let a=new Mt.StringCoder(Number(t.len)).decode((0,Pt.arrayify)(t.data),0)[0];return a=a.replaceAll("\0",""),`${a}`}return`${t.val0}`})}print(){return this.toArray().map((t,e)=>`[log ${e}] ${t}`).join(`
`)}toString(){return this.print()}static debug(t){let e=new V(t);console.log(e.print())}},wt=V;0&&(module.exports={ChangeOutputCollisionError,CoinStatus,CreateTransactionRequest,LogReader,NoWitnessAtIndexError,NoWitnessByOwnerError,Provider,ScriptTransactionRequest,TransactionResponse,TransactionType,calculatePriceWithFactor,coinQuantityfy,getGasUsedFromReceipts,getSignableTransaction,getTransactionId,inputify,outputify,returnZeroScript,transactionRequestify});
